doc/src/manual/constructors.md

```diff
@@ -114,9 +114,11 @@ violate this invariant. Of course, messing around with an object's internals uni
 You (or someone else) can also provide additional outer constructor methods at any later point, but
 once a type is declared, there is no way to add more inner constructor methods. Since outer constructor
 methods can only create objects by calling other constructor methods, ultimately, some inner constructor
-must be called to create an object. This guarantees that all objects of the declared type must come into
-existence by a call to one of the inner constructor methods provided with the type, thereby giving
-some degree of enforcement of a type's invariants.
+must be called to create an object. This means that usually at any given moment all existing objects of
+the declared type have come into existence by a call to one of the inner constructor methods provided
+with the type, thereby giving some degree of enforcement of a type's invariants. This rule does not
+always apply, though: objects may be created without their inner constructors being called when, e.g.,
+[Incomplete Initialization](@ref), [`undef`](@ref) or [`reinterpret`](@ref) are used.
```

Comment:
```suggestion
with the type, thereby giving some degree of enforcement of a type's invariants. This rule does not
always apply, though: objects may be created without their inner constructors being called when, e.g.,
[Incomplete Initialization](@ref), [`undef`](@ref) or [`reinterpret`](@ref) are used.
```
I think you need to clarify:
- the exception only applies to what are called "plain data" types (i.e. those for which `isbitstype` is true, so perhaps link to `isbitstype` documentation?). Any other types will throw a `UndefRefError: access to undefined reference`.
- what specifically you're referring to by incomplete initialization: a type which is incompletely initialized by definition has been constructed through the inner constructor  (which is why I was initially confused by this PR)! What you mean is that a _field_ of an incompletely initialized type may bypass the inner constructor.

Separately, can you also add an `@id` anchor to the "Incomplete Initialization" section header, and `@ref` that (so that it is clear that the section is linked to externally).

Reply:
> the exception only applies to what are called "plain data" types

IMHO this is not correct.

> Any other types will throw

Kind of, yes, but the exception only happens when trying to read the field with the undefined value. It's still possible to do a lot with undefined values:
```julia-repl
julia> struct T{U}
         v::U
         T{U}() where {U <: Any} = new{U}()
       end

julia> T{Vector{Float64}}()
T{Vector{Float64}}(#undef)

julia> mutable struct W{U}
         v::U
         W{U}() where {U <: Any} = new{U}()
       end

julia> w = W{Vector{Float64}}()
W{Vector{Float64}}(#undef)

julia> w.v = Float64[]
Float64[]
```

Ultimately, the issue is that the error you're talking about may be thrown later than the user expects, so they need to know about it. Mentioning plain data would just muddy the waters.

> what specifically you're referring to by incomplete initialization

OK, yes I'll try to clarify this.

> Separately, can you also add an @id anchor to the "Incomplete Initialization" section header, and @ref that (so that it is clear that the section is linked to externally).

OK, but I think you mean "internally", not "externally"?

Reply:
> > the exception only applies to what are called "plain data" types
> 
> IMHO this is not correct.

In what way? Do you have an example?


> > Any other types will throw
> 
> Kind of, yes, but the exception only happens when trying to read the field with the undefined value. It's still possible to do a lot with undefined values:

Sure, but an undefined value isn't an instance of an object. So it is distinct from what we're talking about here.